ambari/src/main/java/io/brooklyn/ambari/AmbariClusterImpl.java

```diff
@@ -131,24 +106,28 @@ public void start(Collection<? extends Location> locations) {
 
         // TODO We could try to prevent setting SERVICE_UP until the services are all intalled
         subscribe(getAttribute(AMBARI_SERVER), AmbariServer.REGISTERED_HOSTS, new RegisteredHostEventListener(this));
+        Iterable<Entity> components = Entities.descendants(this, Predicates.instanceOf(SoftwareProcess.class));
+        EtcHostsManager.setHostsOnMachines(components, getConfig(ETC_HOST_ADDRESS));
     }
 
+
     static final class RegisteredHostEventListener implements SensorEventListener<List<String>> {
+
         private final AmbariCluster entity;
-        
+
         public RegisteredHostEventListener(AmbariCluster entity) {
             this.entity = entity;
         }
-        
         @Override
         public void onEvent(SensorEvent<List<String>> event) {
             List<String> hosts = event.getValue();
-            Integer initialClusterSize = entity.getConfig(INITIAL_SIZE);
+            Integer initialClusterSize = entity.getAttribute(EXPECTED_AGENTS);
             Boolean initialised = entity.getAttribute(CLUSTER_SERVICES_INITIALISE_CALLED);
             if (hosts != null && hosts.size() == initialClusterSize && !Boolean.TRUE.equals(initialised)) {
```

Comment:
Race condition on `initialised`? How much do you care that installServices might be called more than once? Probably better to make `initialise` an `AtomicBoolean`.